test/config/utility.h

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "common/network/address_impl.h"
+
+#include "api/base.pb.h"
+#include "api/bootstrap.pb.h"
+#include "api/cds.pb.h"
+
+namespace Envoy {
+
+class ConfigHelper {
+public:
+  // Set up basic config, using the specified IpVersion for all connections: listeners, upstream,
+  // and admin connections.
+  ConfigHelper(const Network::Address::IpVersion version);
+
+  // Set the upstream ports.  The size of this vector must match the number of socket addresses
+  // across all configured clusters.
+  void setUpstreamPorts(const std::vector<uint32_t>& ports);
+
+  // Set source_address in the bootstrap bind config.
+  void setSourceAddress(const std::string& address_string);
+
+  // Return the bootstrap configuration for dyamic edits.
```

Comment:
Quick question on API here. I kind of like immutability for config as a way of taming complexity. Could we do something similar to GFE, where we allow tests to provide a functional transform of a base config, rather than take a mutable reference to a global config, which might then be subject to arbitrary transform by multiple sites?

Reply:
I'm not convinced arbitrary edits at multiple sites are more or less clear than closures added at multiple sites.  Still given no one is using it it's easy enough to lock it down for now.